File: frontend/src/lib/job-url.ts
/**
 * Normalize user-entered job posting URLs so links work without requiring a scheme in the input.
 * Accepts e.g. "acme.com/careers/123", "//acme.com/jobs", and keeps mailto:/tel: as-is.
 */
export function normalizeJobUrl(raw: string | null | undefined): string | undefined {
  const s = typeof raw === "string" ? raw.trim() : "";
  if (!s) return undefined;
  if (/^https?:\/\//i.test(s)) return s;
  if (s.startsWith("//")) return `https:${s}`;
  if (/^[a-z][a-z0-9+.-]*:\/\//i.test(s)) return s;
  if (/^(mailto|tel|sms):/i.test(s)) return s;
  return `https://${s}`;
}

File: backend/app/utils/job_url.py
"""Normalize job listing URLs for storage and links."""

from __future__ import annotations

import re


def normalize_job_url(raw: str | None) -> str | None:
    if raw is None:
        return None
    s = raw.strip()
    if not s:
        return None
    if re.match(r"^https?://", s, re.I):
        return s
    if s.startswith("//"):
        return f"https:{s}"
    if re.match(r"^[a-z][a-z0-9+.-]*://", s, re.I):
        return s
    if re.match(r"^(mailto|tel|sms):", s, re.I):
        return s
    return f"https://{s}"
